From fa2c9082658c4323577e1706b3293c8fa5e00237 Mon Sep 17 00:00:00 2001
From: Hari Naraayana Desikan Kannan <quic_hnaraaya@quicinc.com>
Date: Fri, 26 Jul 2024 21:36:45 +0530
Subject: [PATCH] Update partner linkId for proper partner link reset.

Currently, when a STA is trying to associate with one link
(say 6 GHz) and due to EAPOL failure (M3 failure) at the STA side,
the STA is not responding leading to deauth. The STA then tries to
associate immediately using another link (say 5 GHz).

At present, we still hold the reference to the sta data until the
maximum inactivity timer (currently 5 seconds) expires, as the STA
may try to associate immediately. When we receive an authentication
request on another link before the expiry of the inactivity timer,
we update the STA's linkid and clear the partner link data as part
of ap_free_sta(). However, the reference is not set to NULL due to
the condition (psta->mld_assoc_link_id == lsta->mld_assoc_link_id)
in set_for_each_partner_link_sta(). As a result, when we receive an
auth request, the sta reference is still available but the data is
cleared after the linkid update, leading to a null pointer deference
at wpa_auth_sm_event().

Fix the issue by updating the linkId for all partner links.
This ensures that the other link context is cleared as part of
set_for_each_partner_link_sta().

Patch-depdendency: r25-hostapd-Add-support-slo-sta-reconnect-with-partner-l.patch

Patch-work: None

Feature: PEER_DELETE

Signed-off-by: Hari Naraayana Desikan Kannan <quic_hnaraaya@quicinc.com>
---
 src/ap/ieee802_11.c |  3 +++
 src/ap/sta_info.c   | 20 ++++++++++++++++++++
 2 files changed, 23 insertions(+)

--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3312,7 +3312,9 @@ static void handle_auth(struct hostapd_d

			ap_sta_set_mld(sta, true);
			sta->mld_assoc_link_id = link_id;
-
+			set_link_id_for_each_partner_link_sta(hapd,
+							      sta,
+							      link_id);
			/*
			 * Set the MLD address as the station address and the
			 * station addresses.
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -912,6 +912,7 @@ struct sta_info * ap_sta_add(struct host
 	ap_sta_remove_in_other_bss(hapd, sta);
 	sta->last_seq_ctrl = WLAN_INVALID_MGMT_SEQ;
 	dl_list_init(&sta->ip6addr);
+	sta->mld_assoc_link_id = -1;
 
 #ifdef CONFIG_TAXONOMY
 	sta_track_claim_taxonomy_info(hapd->iface, addr,
